js/components/select.js

```diff
@@ -0,0 +1,1246 @@
+      this.dispatchEvent(new Event('change'));
+    }
+  }
+
+  /**
+   * A list of currently selected options. It only includes 0 or 1 option because this custom
+   * element doesn’t support multiple selection.
+   * @type {BzOptionElement[]}
+   */
+  get selectedOptions() {
+    return this.options.filter((option) => option.selected);
+  }
+
+  /**
+   * A list of options that are not disabled.
+   * @type {BzOptionElement[]}
+   */
+  get #enabledOptions() {
+    return this.options.filter((option) => option.matches(':not([disabled])'));
+  }
+
+  /**
+   * A list of options that are not disabled or hidden.
+   * @type {BzOptionElement[]}
+   */
+  get #availableOptions() {
+    return this.options.filter((option) => option.matches(':not([disabled], [hidden])'));
+  }
+
+  /**
+   * A reference to the currently active option.
+   * @type {BzOptionElement | undefined}
+   */
+  get #activeOption() {
+    return this.options.find((option) => option.matches('.active'));
+  }
+
+  /**
+   * Make a new option active.
+   * @param {BzOptionElement | undefined} newOption A new option to be an active descendant of the
+   * listbox, or `undefined` to reset the active state.
+   */
+  set #activeOption(newOption) {
+    const currentOption = this.#activeOption;
+
+    if (newOption) {
+      if (currentOption) {
+        currentOption.classList.remove('active');
+
+        if (currentOption.id === `${this.#id}-active-option`) {
+          currentOption.removeAttribute('id');
+        }
+      }
+
+      newOption.id ||= `${this.#id}-active-option`;
+      newOption.classList.add('active');
+      newOption.scrollIntoView();
+      this.#combobox.setAttribute('aria-activedescendant', newOption.id);
+    } else {
+      this.#combobox.removeAttribute('aria-activedescendant');
+    }
+  }
+
+  /**
+   * The currently selected option’s value, or an empty string if no option is selected.
+   * @type {string}
+   */
+  get value() {
+    return this.#selectedOption?.value || '';
+  }
+
+  /**
+   * Select a new option by value.
+   * @param {string} value
+   */
+  set value(value) {
+    this.#selectedOption = this.options.find((option) => option.value === value);
+  }
+
+  /**
+   * Initialize a new `BzSelectElement` instance.
+   */
+  constructor() {
+    super();
+
+    this.#internals = this.attachInternals();
+    this.#id = `select-${Math.random().toString(36).split('.')[1]}`;
+  }
+
+  /**
+   * Called when the element is added to the document. Initialize the internals including the shadow
+   * DOM and event handers.
+   */
+  connectedCallback() {
+    this.#label =
+      (this.hasAttribute('aria-labelledby')
+        ? document
+            .getElementById(this.getAttribute('aria-labelledby'))
+            ?.textContent?.trim()
+            .replace(/:$/, '')
+        : this.getAttribute('aria-label')) || this.getAttribute('name');
+
+    this.attachShadow({ mode: 'open' }).innerHTML = `
+      <style>
+        :host {
```

Comment:
Would the css be better in a separate file outside of the js code or was there a reason it needed to be here?

Reply:
This is the way custom components are encapsulated. I personally don’t like this, but CSS is often written in JS. Should be fine.